sycl/source/detail/pi.cpp

```diff
@@ -309,14 +311,316 @@ bool trace(TraceLevel Level) {
+  device_filter_list *FilterList = SYCLConfig<SYCL_DEVICE_FILTER>::get();
+  if (!FilterList)
+    return;
+
+  backend Backend = Plugin.getBackend();
+  int InsertIDx = 0;
+
+  for (RT::PiDevice Device : PiDevices) {
+    RT::PiDeviceType PiDevType;
+    Plugin.call<PiApiKind::piDeviceGetInfo>(Device, PI_DEVICE_INFO_TYPE,
+                                            sizeof(RT::PiDeviceType),
+                                            &PiDevType, nullptr);
+    // Assumption here is that there is 1-to-1 mapping between PiDevType and
+    // Sycl device type for GPU, CPU, and ACC.
+    info::device_type DeviceType = pi::cast<info::device_type>(PiDevType);
+
+    for (const device_filter &Filter : FilterList->get()) {
+      backend FilterBackend = Filter.Backend;
+      // First, match the backend entry
+      if (FilterBackend == Backend || FilterBackend == backend::all) {
+        info::device_type FilterDevType = Filter.DeviceType;
+        // Next, match the device_type entry
+        if (FilterDevType == info::device_type::all) {
+          // Last, match the device_num entry
+          if (!Filter.HasDeviceNum || DeviceNum == Filter.DeviceNum) {
+            PiDevices[InsertIDx++] = Device;
+            break;
+          }
+        } else if (FilterDevType == DeviceType) {
+          if (!Filter.HasDeviceNum || DeviceNum == Filter.DeviceNum) {
+            PiDevices[InsertIDx++] = Device;
+            break;
+          }
+        }
+      }
+    }
+    DeviceNum++;
+  }
+  PiDevices.resize(InsertIDx);
+}
+
+// Fill up the platform cache and device cache for the given Plugin.
+// Thi<s is a one time process when a plugin is constructed.
+void fillPlatformAndDeviceCache(plugin &Plugin) {
+  pi_uint32 NumPlatforms = 0;
+  // Move to the next plugin if the plugin fails to initialize.
+  // This way platforms from other plugins get a chance to be discovered.
+  if (Plugin.call_nocheck<PiApiKind::piPlatformsGet>(
+          0, nullptr, &NumPlatforms) != PI_SUCCESS)
+    return;
+
+  if (NumPlatforms) {
+    vector_class<RT::PiPlatform> PiPlatforms(NumPlatforms);
+    if (Plugin.call_nocheck<PiApiKind::piPlatformsGet>(
+            NumPlatforms, PiPlatforms.data(), nullptr) != PI_SUCCESS)
+      return;
+
+    std::vector<PlatformImplPtr> &PlatformCache =
+        GlobalHandler::instance().getPlatformCache();
+    std::vector<DeviceImplPtr> &DeviceCache =
+        GlobalHandler::instance().getDeviceCache();
+
+    int DeviceNum = 0;
+    for (const auto &PiPlatform : PiPlatforms) {
+      PlatformImplPtr PlatformImpl =
+          std::make_shared<platform_impl>(PiPlatform, Plugin);
+      {
+        const std::lock_guard<std::mutex> Guard(
+            GlobalHandler::instance().getPlatformMapMutex());
+
+        PlatformCache.emplace_back(PlatformImpl);
+      }
+      platform Platform = detail::createSyclObjFromImpl<platform>(PlatformImpl);
+
+      if (IsBannedPlatform(Platform))
+        continue;
```

Comment:
Shouldn't we check this before adding to the `PlatformCache`?

Reply:
Correct. I moved caching platform_impl after this.